staging/src/k8s.io/apimachinery/pkg/util/httpstream/wsstream/conn.go

```diff
@@ -168,15 +187,47 @@ func (conn *Conn) SetIdleTimeout(duration time.Duration) {
 	conn.timeout = duration
 }
 
+// SetWriteDeadline sets a timeout on writing to the websocket connection. The
+// passed "duration" identifies how far into the future the write must complete
+// by before the timeout fires.
+func (conn *Conn) SetWriteDeadline(duration time.Duration) {
+	conn.ws.SetWriteDeadline(time.Now().Add(duration)) //nolint:errcheck
+}
+
 // Open the connection and create channels for reading and writing. It returns
 // the selected subprotocol, a slice of channels and an error.
 func (conn *Conn) Open(w http.ResponseWriter, req *http.Request) (string, []io.ReadWriteCloser, error) {
+	// serverCompleted is channel indicating server finished before it became ready.
+	serverCompleted := make(chan struct{})
+	// Ensure panic in spawned goroutine is propagated into the parent goroutine.
+	panicChan := make(chan any, 1)
 	go func() {
-		defer runtime.HandleCrash()
-		defer conn.Close()
+		// If websocket server returns, propagate panic if necessary.
+		defer func() {
+			if p := recover(); p != nil {
+				panicChan <- p
+			}
+			close(serverCompleted)
+		}()
 		websocket.Server{Handshake: conn.handshake, Handler: conn.handle}.ServeHTTP(w, req)
 	}()
-	<-conn.ready
+
+	select {
+	case <-conn.ready:
+		klog.V(8).Infof("websocket server initialized--serving")
+	case <-serverCompleted:
+		// websocket server returned before completing initialization; cleanup and return error.
+		for _, c := range conn.channels {
+			c.w.Close() //nolint:errcheck
+		}
+		if conn.ws != nil {
+			conn.ws.Close() //nolint:errcheck
+		}
```

Comment:
put this into a private function named `closeNonThreadSafe()` or something and call it from conn#Close() and here

Reply:
Done